### PR TITLE
[#9] [Integrate] login screen

### LINKED ISSUE #9: [Integrate] As a user, I can login with email and password
## Why

App must call backend API to perform login

## Acceptance Criteria

- Make post request to `https://survey-api.nimblehq.co/api/v1/oauth/token` and handle response
- Display a loading indicator while request is ongoing
- If successful, display a success message
  - Navigation to loading will be handled in #14 
- If unsuccessful, display a toast with the error message from API

## Resources

- [API](https://documenter.getpostman.com/view/11835486/UVsPNPfU#deb6f64d-60a4-434b-9c83-215ad8a07568)

### PR DESCRIPTION
**Pull request by lydiasama**
- Closes https://github.com/nimblehq/flutter-ic-lydia-ryan/issues/9

## What happened 👀

- Added `LoginViewModel` to call `LoginUseCase` after clicking on `Login` button.
- Integrated to get text from input fields and pass them to `LoginUseCase`.
- Added [email_validator](https://pub.dev/packages/email_validator) library to validate email.
- Added [flutter_riverpod](https://pub.dev/packages/flutter_riverpod/example) library to manage state.
- Added [fluttertoast](https://pub.dev/packages/fluttertoast) to show Toast message.
- Added `BaseViewModelState` instead of create `LoginState` to avoid boilerplate code for future screens.
- Added `CircleLoadingIndicator` widget.

## Insight 📝

- Fix dependency injection: We didn't call `configureDependencies()` function on `main.dart` yet. So the dependency injection will not work. 😂 
- Need to upgrade the Flutter SDK version to `3.7.11` for compatible SDK version with `fluttertoast:8.2.1`.
- Let's use `@GenerateNiceMocks` instead of `@GenerateMocks` as all method of mocked class is stubbed even though you didn’t put the code explicitly in the unit test. And it's the new approach from Mockito. [Ref here](https://levelup.gitconnected.com/flutter-test-mockito-generatemocks-vs-generatenicemocks-3dbadfe14afc).
 
## Proof Of Work 📹

After inputting email and password and clicking on `Login` button:

```
flutter: *** Response ***
flutter: uri: https://survey-api.nimblehq.co/api/v1/oauth/token
flutter: statusCode: 200
flutter: headers:
flutter:  connection: keep-alive
flutter:  cache-control: no-store
flutter:  date: Wed, 12 Apr 2023 08:10:39 GMT
flutter:  vary: Accept-Encoding, Origin
flutter:  content-encoding: gzip
flutter:  referrer-policy: strict-origin-when-cross-origin
flutter:  report-to: {"endpoints":[{"url":"https:\/\/a.nel.cloudflare.com\/report\/v3?s=01u0YXYQ0sL1BqIGdHh1ImsUv2LEs4yPMU2%2F%2B3ayswafHw%2FRIdKX0uVjPzD9rfK4L1WEBP9o6SZVKgtK4NLrRlODeda6Kh26mJ%2BKmyE2c6HmSq1Izx4i7qMJD3jvhS4TcTp3p9Wljh1M"}],"group":"cf-nel","max_age":604800}
flutter:  cf-cache-status: DYNAMIC
flutter:  x-permitted-cross-domain-policies: none
flutter:  content-type: application/json; charset=utf-8
flutter:  x-xss-protection: 0
flutter:  server: cloudflare
flutter:  x-request-id: fb19b29e-b557-47bd-88ca-83dde8e1690d
flutter:  alt-svc: h3=":443"; ma=86400, h3-29=":443"; ma=86400
flutter:  content-length: 218
flutter:  nel: {"success_fraction":0,"report_to":"cf-nel","max_age":604800}
flutter:  x-download-options: noopen
flutter:  cf-ray: 7b69fe7329a426cd-BKK
flutter:  x-runtime: 0.460254
flutter:  etag: W/"783fe4a958e84b74246c663c08d2ab9d"
flutter:  via: 1.1 vegur
flutter:  x-frame-options: SAMEORIGIN
flutter:  x-content-type-options: nosniff
flutter: Response Text:
flutter: {"data":{"id":"1022","type":"token","attributes":{"access_token":"7Dmz5yiYozY6r9YC25M7XQc_JlvN01xa3YgLfBYCimU","token_type":"Bearer","expires_in":7200,"refresh_token":"cGQLtv1Wk0yTTIxMlcLUdz_No96D1U2exek5644ue3A","created_at":1681287039}}}
```
<details>
<summary> iOS </summary>

https://user-images.githubusercontent.com/12026942/232672665-f0a09bfa-149e-4d4e-af19-2f3b3468d752.mp4

</details>

<details>
<summary> Android </summary>

[android.webm](https://user-images.githubusercontent.com/12026942/232672724-3d6f6bf1-72ab-4467-b44e-136860fa13c6.webm)

</details>

### CODE REVIEW

File: lib/screens/login/login_view_model.dart [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+      );
+      if (result is Success) {
+        state = const BaseViewModelState.success();
+      } else {
+        state =
+            BaseViewModelState.apiError((result as Failed).getErrorMessage());
+      }
+    } else {
+      state = const BaseViewModelState.invalidInputsError();
+    }
```
Comment (markgravity) [inline lib/screens/login/login_view_model.dart]: What do you think about cleaning up them a bit? 
```swift
    if (!_validateInputs(email, password)) {
     state = const BaseViewModelState.invalidInputsError();
     return;
    }
    
    Result<void> result = await _loginUseCase.call(
        LoginInput(
          email: email,
          password: password,
        ),
      );
      if (result is Success) {
        state = const BaseViewModelState.success();
      } else {
        state =
            BaseViewModelState.apiError((result as Failed).getErrorMessage());
      }
```

File: lib/screens/login/login_view_model.dart [hunk 2]
```diff
@@ -0,0 +1,38 @@
+import 'package:email_validator/email_validator.dart';
+import 'package:flutter_riverpod/flutter_riverpod.dart';
+import 'package:lydiaryanfluttersurvey/base/base_view_model_state.dart';
+import 'package:lydiaryanfluttersurvey/usecases/login_use_case.dart';
+
+import '../../usecases/base/base_use_case.dart';
+
+class LoginViewModel extends StateNotifier<BaseViewModelState> {
```
Comment (markgravity) [inline lib/screens/login/login_view_model.dart:8]: Please write unit test for this view model 🙏

**Review by markgravity — COMMENTED** [review]
We are missing these ACs
- Display a loading indicator while request is ongoing
- If successful, display a success message
- If unsuccessful, display a toast with the error message from API

**Comment by lydiasama** [comment]
@markgravity Do you mean this [loading](https://github.com/orgs/nimblehq/projects/39/views/3?pane=issue&itemId=21306836)? Or is there any other loading indicator in Figma? 🤔

**Comment by ryan-conway** [comment]
> Do you mean this [loading](https://github.com/orgs/nimblehq/projects/39/views/3?pane=issue&itemId=21306836)? Or is there any other loading indicator in Figma? 🤔

@lydiasama I think we should show a loading indicator while executing the login request. [This ticket](https://github.com/nimblehq/flutter-ic-lydia-ryan/issues/14) is to show the loading state when fetching surveys on the home screen 😇  Perhaps replace the button with a spinning indicator? There isn't an example on Figma 🥲

**Comment by lydiasama** [comment]
@ryan-conway It's just a simple circle loading, right? Let me add it. I just don't see any design in figma. 🙈

File: lib/usecases/base/base_use_case.dart [hunk 3]
```diff
@@ -1,3 +1,5 @@
+import '../../api/exception/network_exceptions.dart';
```
Comment (ryan-conway) [inline lib/usecases/base/base_use_case.dart:1]: Why do we need this import? 🤔

**Review by ryan-conway — COMMENTED** [review]
Just one comment 😄

File: lib/usecases/base/base_use_case.dart (hunk 3 above)
Comment (lydiasama) [reply lib/usecases/base/base_use_case.dart:1]: @ryan-conway It's used by `use_case_result.dart`. I know it's weird, but when importing it by the IDE, it was added here. I'm not sure why. I also tried to move it to `use_case_result` too. But it doesn't work.

**Review by ryan-conway — APPROVED** [review]
🚀

File: lib/usecases/base/base_use_case.dart (hunk 3 above)
Comment (lydiasama) [reply lib/usecases/base/base_use_case.dart:1]: @ryan-conway You can try to fix it on your end. 😄

**Review by markgravity — APPROVED** [review]

**Pull request merged**